Retrieve previously generated résumé items

MithishR opened the pull request:
<!-- 
Thank you for contributing! Please fill out this template to help us review your PR.
-->

## 📝 Description

This PR implements a comprehensive résumé retrieval and regeneration system that allows users to efficiently manage previously generated résumé items from the database, thus meeting the requirement for milestone 1. The system provides intelligent caching, selective regeneration, and improved user experience with cleaner output formatting. 

**Key Changes:**
- Implemented smart caching with options to view, regenerate, or skip résumé generation
- Improved résumé output formatting for better readability
- Enhanced error handling in database operations with proper validation and logging

**IMPORTANT: Please merge #183 before merging this!!!**

**Closes:** #172 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [x] ⚡ Performance improvement

---

## 🧪 Testing

**Automated testing:** Unit tests were added. Run ` pytest src/tests/backend_test/test_resume_retrieval.py -v`. All tests should pass. 

**Manual testing:**
Run `python src/backend/analysis/analyze.py <zip-file>` and select any option if existing analysis is detected. Let the analysis happen.
Once the analysis occurs, there will be 3 prompts displayed if resume items were already in the database.
  1. View existing résumé items (and generate missing ones)
  2. Regenerate all résumé items (overwrite existing)
  3. Skip résumé generation

For (1): Should display: `✓ Retrieved X cached résumé item(s) from database`. `X` resume items should be displayed in the terminal and you can verify the database to see if it's the same items being displayed.

For (2): Should display: `Regenerating all X résumé item(s)...`. After this, `X` resume items should be displayed in the terminal and you can see that the database contains new resume items with the most recent date and time, and the previous resume items should be deleted.

For (3): No resume items will be displayed and nothing is deleted/overwritten. Program should move gracefully to the next step (Saving report to the JSON file). 


---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [ ] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [ ] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

<!-- Add your screenshots here -->

</details>

anshr18 reviewed (approved):
Approved 
Tested the updated résumé flow via `analyze.py` with existing résumé items in the DB and the behavior matches the description:

- Option **(1) View existing résumé items** correctly retrieves cached entries, prints `✓ Retrieved X cached résumé item(s) from database`, and shows the expected items without duplicating or corrupting data.
- Option **(2) Regenerate all résumé items** overwrites previous entries, shows `Regenerating all X résumé item(s)...`, and the DB reflects only the new items with updated timestamps.
- Option **(3) Skip résumé generation** cleanly bypasses résumé output and continues to the next steps (JSON report) without side effects.

Unit tests in `test_resume_retrieval.py` pass as well.


<img width="571" height="757" alt="Screenshot 2025-12-04 at 12 51 19 AM" src="https://github.com/user-attachments/assets/a31f2286-737e-4efc-8cda-e0c63312d061" />

<img width="1698" height="1548" alt="image" src="https://github.com/user-attachments/assets/3e3a9719-c0ae-421d-b9df-c350ba5cdac9" />

<img width="1193" height="437" alt="Screenshot 2025-12-04 at 12 48 04 AM" src="https://github.com/user-attachments/assets/544a995d-0f44-49f8-99fd-5b8b2ae85f7c" />

HarjotSahota reviewed (changes requested):
Great work overall the resume retrieval and regeneration flow works smoothly, and I was able to run through all three branches (view, regenerate, skip) without issues. Unit tests all passed cleanly. I did encounter one reproducible edge-case bug during manual testing: (this involves your previous pr with https://github.com/COSC-499-W2025/capstone-project-team-6/pull/183)


During testing, I was able to consistently reproduce the following error:
```
 UnboundLocalError: local variable 'newly_generated' referenced before assignment 
```

**Reproduce Bug:**
1. Run analyze.py on a project with 1 resume item.
2. Generate resume items once.
3. Re-run the script.
4. Select:
    - 1 (delete 0 older analyses)
    - 1 (view existing résumé items)

Program crashes immediately after displaying the cached résumé item.

**Cause:**
`newly_generated` is only defined inside the branch that generates new resume items.
When the user only views existing items, that variable never gets assigned but is still printed afterward. So the variable is referenced even though the code path never created it.

**optional fix:**
Add a default initialization:
```
newly_generated = 0
```
And guard the print:
```
if projects_needing_resume:
    print(f" Successfully generated and stored {newly_generated}/{len(projects_needing_resume)} new resume item(s)")
else:
    print(" No new resume items to generate.")
```
This avoids the crash entirely and matches expected behavior in all three flows.

all tests pass, once this bug is fixed I am happy to approve! 

<img width="1152" height="485" alt="Screenshot 2025-12-05 at 5 18 22 PM" src="https://github.com/user-attachments/assets/c55c6237-ad0a-4284-845e-52e181aaf7dd" />

MithishR commented:
@HarjotSahota thank you for your feedback and for catching that error-not sure how I missed that. I have addressed and implemented your changes, please have a look

HarjotSahota reviewed (approved):
Great work on fixing the bug, everything works as intended! 

Approving this pr!
